OpenSim/Simulation/Model/ContactGeometry.cpp

```diff
@@ -81,90 +74,95 @@ void ContactGeometry::constructProperties()
     constructProperty_color(defaultColor);
 }
 
-const Vec3& ContactGeometry::getLocation() const
+void ContactGeometry::constructConnectors()
 {
-    return get_location();
+    constructConnector<PhysicalFrame>("frame");
 }
 
+const Vec3& ContactGeometry::getLocation() const
+{ return get_location(); }
+
 void ContactGeometry::setLocation(const Vec3& location)
-{
-    set_location(location);
-}
+{ set_location(location); }
 
 const Vec3& ContactGeometry::getOrientation() const
-{
-    return get_orientation();
-}
+{ return get_orientation(); }
 
 void ContactGeometry::setOrientation(const Vec3& orientation)
-{
-    set_orientation(orientation);
-}
-
-SimTK::Transform ContactGeometry::getTransform()
-{
-    return SimTK::Transform(Rotation(SimTK::BodyRotationSequence,
-        get_orientation()[0], SimTK::XAxis,
-        get_orientation()[1], SimTK::YAxis,
-        get_orientation()[2], SimTK::ZAxis), get_location());
-}
-
-PhysicalFrame& ContactGeometry::updBody()
-{
-    return *_body;
-}
+{ set_orientation(orientation); }
 
-const PhysicalFrame& ContactGeometry::getBody() const
+SimTK::Transform ContactGeometry::findTransformInBaseFrame() const
 {
-    return *_body;
+    // B: base Frame (Body)
+    // F: PhysicalFrame specified in the connector for this object.
+    // P: the frame that is defined (relative to F) by the location
```

Comment:
I still do not understand why it is the ContactGeometry's responsibility to provide X_BP. If the user only provides P in F then return that. If the ContactForce implementation requires a transform wr.t. MobilizedBody it can always ask any F for its BaseFrame. I would prefer that in order to simplify ContactGeometry.